LibGfx+LibWeb: Move EXIF orientation to rendering and implement image-orientation

JohnyTheCarrot opened the pull request:
This PR aims to implement support for EXIF orientation and the `image-orientation` CSS property.
Previous existing implementations for EXIF orientation performed orientation while decoding, which makes implementing the `image-orientation` property difficult.

This PR instead:
1. Adds the exif orientation as a member to the bitmap, allowing us to read it out while rendering.
2. Performs the exif orientations while rendering, leaving the bitmap data alone.

Adding support for other image formats should be as simple as passing the EXIF orientation to the bitmap once it is retrieved from the EXIF data in the decoder.

![output](https://github.com/user-attachments/assets/0c803cd4-5754-448c-a7ea-8906648b5f86)

See the following links for prior discussions.
https://github.com/LadybirdBrowser/ladybird/issues/2065
https://github.com/LadybirdBrowser/ladybird/pull/2122#issuecomment-2458172026

JohnyTheCarrot commented:
I tried to add [this test](https://wpt.live/css/css-images/image-orientation/image-orientation-from-image.html) but it. strangely enough, fails.

The expected screenshot and the actual screenshot match almost exactly outside of some small bit of artifacting.
A diff highlighted image comparison is included below.
How do we proceed? Is this something I did, or is this an existing issue?

![image](https://github.com/user-attachments/assets/db06410b-9c1c-43d3-84dc-b8b0ac80c151)

AtkinsSJ commented on Libraries/LibGfx/Bitmap.cpp:
```diff
@@ -20,6 +20,34 @@ struct BackingStore {
     size_t size_in_bytes { 0 };
 };
 
+[[nodiscard]]
+IntSize oriented_size(IntSize size, Gfx::ExifOrientation orientation);
+
+IntSize oriented_size(IntSize size, Gfx::ExifOrientation orientation)
```
Nit: Combine these together and mark the function as `static`.

AtkinsSJ commented on Libraries/LibGfx/ImmutableBitmap.cpp:
```diff
@@ -31,6 +31,24 @@ int ImmutableBitmap::height() const
+        return width();
+
+    auto const &bitmap = m_impl->source.get<NonnullRefPtr<Gfx::Bitmap>>();
+    return bitmap->width_considering_exif();
```
You can simplify this a little by using `get_pointer` instead of `get`. It'll return a pointer to the value if it has one, and null if it doesn't, so it's equivalent to `has` and `get`, but only has to do the work once.

AtkinsSJ commented on Libraries/LibWeb/Painting/DisplayListPlayerSkia.cpp:
```diff
@@ -133,13 +134,86 @@ void DisplayListPlayerSkia::draw_painting_surface(DrawPaintingSurface const& com
 
+void apply_exif_orientation(Gfx::PaintingSurface &surface, Gfx::ExifOrientation orientation, Gfx::IntRect &dst_rect);
+
+void apply_exif_orientation(Gfx::PaintingSurface &surface, Gfx::ExifOrientation orientation, Gfx::IntRect &dst_rect) {
```
Same nit about `static`.

AtkinsSJ commented on Libraries/LibGfx/Bitmap.cpp:
```diff
@@ -40,11 +40,13 @@ IntSize oriented_size(IntSize size, Gfx::ExifOrientation orientation)
 }
 
-int Bitmap::width_considering_exif() const {
+int Bitmap::width_considering_exif() const
```
Please move those changes to the first commit, so each commit is correct by itself. [This video](https://www.youtube.com/watch?v=ElRzTuYln0M) explains how to do that pretty well if you're not familiar with it.

If you want, there's a [git commit hook](https://github.com/LadybirdBrowser/ladybird/blob/master/CONTRIBUTING.md#commit-hooks) available, which runs clang-format automatically each time you try to commit, and makes sure your commit messages match our rules too. (It's quite handy. 😅)

AtkinsSJ commented on Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:
```diff
@@ -87,11 +92,27 @@ ErrorOr<void> JPEGLoadingContext::decode()
+        if (marker->marker == jpeg_app1 && marker->data_length >= 6) {
+            JOCTET expected[6] = { 'E', 'x', 'i', 'f', 0, 0 };
+
+            if (memcmp(marker->data, expected, 6) != 0) {
```
Generally we try and avoid things like `memcmp()` and use something more type-safe instead. However, this looks correct to me, and image decoding is not at all my area so maybe there isn't a nicer way to do it at the moment.

LucasChollet replied:
```c++
static constexpr auto expected = "Exif\0\0"sv;
auto signature = StringView { marker->data, 6 };
if (signature != expected) {
	[...]
}
```
That should work!

AtkinsSJ reviewed (changes requested):
Hi! This is very nice for a first PR. :^)

A couple of small things, but nothing major. A test would be good of course, but I know that's already been discussed on Discord. *EDIT: And literally in a comment right above this.* 😆

kalenikaliaksandr commented on Libraries/LibWeb/Painting/DisplayListPlayerSkia.cpp:
```diff
@@ -133,13 +134,89 @@ void DisplayListPlayerSkia::draw_painting_surface(DrawPaintingSurface const& com
+        canvas.translate(x * 2 + dst_rect.width() / 2, 0);
+        canvas.scale(-1, 1);
+        canvas.translate(-dst_rect.width() / 2, 0);
+        return;
+    case Gfx::ExifOrientation::Rotate180:
+        canvas.translate(dst_rect.width(), dst_rect.height());
+        canvas.rotate(180, x, y);
+        return;
+    case Gfx::ExifOrientation::FlipVertically:
+        canvas.translate(0, y * 2 + dst_rect.height() / 2);
+        canvas.scale(1, -1);
+        canvas.translate(0, -dst_rect.height() / 2);
+        return;
+    case Gfx::ExifOrientation::Rotate90ClockwiseThenFlipHorizontally:
+        canvas.translate(x * 2 + dst_rect.height() / 2, 0);
+        canvas.scale(-1, 1);
+        // correct for scaling: -dst_rect.height() / 2 plus correct for rotation: dst_rect.height() becomes dst_rect.height() / 2
+        canvas.translate(dst_rect.height() / 2, 0);
+        canvas.rotate(90, x, y);
+        return;
+    case Gfx::ExifOrientation::Rotate90Clockwise:
+        canvas.translate(dst_rect.height(), 0);
+        canvas.rotate(90, x, y);
+        return;
+    case Gfx::ExifOrientation::FlipHorizontallyThenRotate90Clockwise:
+        canvas.translate(dst_rect.height(), 0);
+        canvas.rotate(90, x, y);
+
+        canvas.translate(x * 2 + dst_rect.width() / 2, 0);
+        canvas.scale(-1, 1);
+        canvas.translate(-dst_rect.width() / 2, 0);
+        return;
+    case Gfx::ExifOrientation::Rotate90CounterClockwise:
+        canvas.translate(0, dst_rect.width());
+        canvas.rotate(-90, x, y);
+        return;
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
```
bringing logic that maps EXIF orientation to canvas transforms is a layering violation. the purpose of the display list is to serve as a data structure that contains sufficient information to perform painting while omitting all irrelevant details, such as underlying image formats. I suggest saving the transformation matrix in the `DrawScaledImmutableBitmap` struct instead.

JohnyTheCarrot replied:
Not entirely sure what you're suggesting.
Moving the matrix to DrawScaledImmutableBitmap would make the caller responsible for constructing the matrix based off the exif orientation, no?
What do you suggest I do instead?

kalenikaliaksandr replied:
my suggestion is to construct transformation matrux while recording display list and then save it in DrawScaledImmutableBitmap for use while painting. that would be better because underlying format knowledge won't leak into painting implementation.

kalenikaliaksandr commented on Libraries/LibGfx/Bitmap.h:
```diff
@@ -160,6 +164,7 @@ class Bitmap : public RefCounted<Bitmap> {
     AlphaType m_alpha_type { AlphaType::Premultiplied };
     Core::AnonymousBuffer m_buffer;
     Function<void()> m_destruction_callback;
+    ExifOrientation m_exif_orientation;
```
instead of making Gfx::Bitmap aware of EXIF orientation, can we construct bitmap with rotated image directly? this way, hacks like `width_considering_exif()` and `height_considering_exif()` would no longer be necessary.

JohnyTheCarrot replied:
Not to my understanding, we need to retain the original bitmap because a the CSS property `image-orientation` could disable EXIF orientations entirely.

kalenikaliaksandr replied:
ok, you are right, though I assume `image-orientation` property is unlikely to change frequently, so it might be reasonable to store it with orientation applied and redo decoding whenever `image-orientation` changes.

JohnyTheCarrot commented:
Ready for another review.
I do realize my SkMatrix -> Matrix3x3 -> SkMatrix shenanigans may raise eyebrows but it seemed that there was a component using Command.h that did not add Skia to its includes.
If this is a problem please do advise on how I could solve it.

justfitting commented:
The things you shared are great. I want to share my happiness with you. I found a fun website(https://aikungfugenerator.com), where I can turn my photos into action videos. It’s perfect for social media and gets lots of attention! Compared to the Sprunki game websites ([Sprunki Phase](https://sprunkiphase.online), [Corruptbox](https://corruptboxsprunki.online), [Sprunki 2025](https://sprunki2025.com)), which are more about music, AI Kungfu is way more dynamic and entertaining.

kalenikaliaksandr commented on Libraries/LibWeb/HTML/HTMLImageElement.cpp:
```diff
@@ -170,22 +170,38 @@ bool HTMLImageElement::is_image_available() const
         return image_data->intrinsic_width();
+    }
+
     return {};
```
changing HTMLImageElement to look into bitmap directly instead of going trough DecodedImageData seems wrong because there are also HTMLObjectElement and HTMLInputElement which are not aware of this change. can we instead update DecodedImageData to get size with orientation applied?

kalenikaliaksandr commented on Libraries/LibWeb/Painting/Command.h:
```diff
@@ -75,6 +77,7 @@ struct DrawScaledImmutableBitmap {
     NonnullRefPtr<Gfx::ImmutableBitmap> bitmap;
     Gfx::IntRect src_rect;
     Gfx::ScalingMode scaling_mode;
+    Gfx::Matrix3x3<float> transmat;
```
let's use `transformation_matrix` name here.

kalenikaliaksandr commented on Libraries/LibWeb/Painting/DisplayListRecorder.cpp:
```diff
@@ -6,6 +6,7 @@
 
 #include <LibWeb/Painting/DisplayListRecorder.h>
 #include <LibWeb/Painting/ShadowPainting.h>
+#include <core/SkMatrix.h>
```
we don't use Skia directly in LibWeb outside of DisplayListPlayerSkia. let's use matrix class from LibGfx instead.

JohnyTheCarrot commented:
Ready for review.
Sorry for the large amount of changes, this PR has grown significantly.
My changes since the last review implement image orientation for everywhere it's supported.
The only thing I did not get to is ignoring `image-orientation: none` for cross site images.

InvalidUsernameException reviewed (commented):
Partial, non-maintainer review. So take it with a grain of salt.

JohnyTheCarrot commented on Tests/LibWeb/Screenshot/input/exif-orientation-png.html:
Good catch! How the hell did that happen?

JohnyTheCarrot commented on Libraries/LibGfx/ImageOrientation.cpp:
```diff
@@ -0,0 +1,97 @@
+    }
+}
+
+ImageOrientation to_gfx_image_orientation(Web::CSS::ImageOrientation orientation)
+{
+    return static_cast<ImageOrientation>(orientation);
+}
```
So you'd advise moving this function to LibWeb/CSS?

InvalidUsernameException replied:
That would seem like a more natural place to me, yes.

JohnyTheCarrot commented on Libraries/LibWeb/Painting/DisplayListRecorder.cpp:
```diff
@@ -194,26 +194,61 @@ void DisplayListRecorder::draw_painting_surface(Gfx::IntRect const& dst_rect, No
+
+    auto clip_rect_oriented = clip_rect;
+    auto const exif_orientation = bitmap.get_exif_orientation();
+    if (image_orientation == Gfx::ImageOrientation::FromExif) {
```
You are correct... ish. The dst_rect is already correct at this point, you are right on that front.
However, we are applying a rotation to the canvas, which "invalidates" the dst_rect in that the dst_rect now refers to a rotated canvas. In order to fix this we need to swap the width and height so we're back in the desired aspect ratio.

That said, your review comment inspired me to clarify this in code.
I've also realized this is a Skia specific quirk, and should probably not be handled in the recorder in case some other renderer is ever installed.

JohnyTheCarrot commented:
@InvalidUsernameException 's review comments have been addressed

AtkinsSJ commented on Libraries/LibGfx/ImageFormats/PNGLoader.cpp:
```diff
@@ -259,7 +234,7 @@ ErrorOr<size_t> PNGLoadingContext::read_frames(png_structp png_ptr, png_infop in
         return frame_bitmap;
     };
-
+    
```
The linter won't like this trailing whitespace.

AtkinsSJ commented on Libraries/LibGfx/ImageOrientation.h:
```diff
@@ -0,0 +1,101 @@
+    Rotate90CounterClockwise = 8,
+};
+
+[[nodiscard]]
+bool is_valid_exif_orientation(u32 orientation);
```
`[[nodiscard]]` on the same line.

AtkinsSJ commented on Libraries/LibGfx/ImageOrientation.h:
```diff
@@ -0,0 +1,101 @@
+bool is_valid_exif_orientation(u32 orientation);
+
+template<typename T>
+[[nodiscard]]
```
Same

AtkinsSJ commented on Libraries/LibGfx/ImmutableBitmap.h:
```diff
@@ -41,6 +41,9 @@ class ImmutableBitmap final : public AtomicRefCounted<ImmutableBitmap> {
 
     RefPtr<Bitmap const> bitmap() const;
 
+    [[nodiscard]]
```
Same

AtkinsSJ commented on Libraries/LibGfx/PainterSkia.cpp:
```diff
@@ -142,6 +142,21 @@ void PainterSkia::fill_rect(Gfx::FloatRect const& rect, Color color)
     });
 }
 
+static SkMatrix to_skia_matrix(Gfx::AffineTransform const& affine_transform)
```
You have this here and in `DisplayListPlayerSkia.cpp` - move it to `SkiaUtils` instead.

AtkinsSJ commented on Libraries/LibWeb/Painting/DisplayListRecorder.cpp:
```diff
@@ -194,26 +194,57 @@ void DisplayListRecorder::draw_painting_surface(Gfx::IntRect const& dst_rect, No
 }
 
+[[nodiscard]]
+static bool is_rect_correction_necessary(Gfx::ExifOrientation exif_orientation)
```
Nit: This feels like it belongs in `ImageOrientation`, as something like `exif_orientation_affects_image_size()`.

AtkinsSJ commented on Libraries/LibGfx/ImageOrientation.h:
```diff
@@ -11,8 +11,17 @@
 #include <LibGfx/Rect.h>
 #include <cmath>
 
+namespace Web::CSS {
+enum class ImageOrientation : u8;
+}
```
Why is this needed?

AtkinsSJ commented on Libraries/LibWeb/CSS/ComputedValues.h:
```diff
@@ -345,24 +345,6 @@ struct BorderRadiusData {
     CSS::LengthPercentage vertical_radius { InitialValues::border_radius() };
 };
 
-// FIXME: Find a better place for this helper.
-inline Gfx::ScalingMode to_gfx_scaling_mode(CSS::ImageRendering css_value, Gfx::IntRect source, Gfx::IntRect target)
```
Moving this isn't part of the EXIF change, so it'd be good to put it in its own commit.

AtkinsSJ commented on Libraries/LibGfx/ImageOrientation.h:
```diff
@@ -0,0 +1,101 @@
+        break;
+    case Gfx::ExifOrientation::Rotate180:
+        matrix.set_translation(dst_rect.width(), dst_rect.height());
+        matrix.rotate_radians(M_PI);
```
Instead of the `<cmath>` constants, use the ones in `<AK/Math.h>`.

AtkinsSJ commented on Tests/LibGfx/TestImageDecoder.cpp:
```diff
@@ -352,21 +352,6 @@ TEST_CASE(test_png)
     TRY_OR_FAIL(expect_single_frame(*plugin_decoder));
 }
 
-TEST_CASE(test_exif)
```
Is there some way we can still test the EXIF decoding?

AtkinsSJ commented on Libraries/LibWeb/HTML/Canvas/CanvasDrawImage.cpp:
```diff
@@ -62,6 +66,38 @@ static void default_source_size(CanvasImageSource const& image, float& source_wi
 
+Gfx::ImageOrientation get_image_orientation_from_canvas_source(CanvasImageSource const& source)
+{
+    auto const image_orientation = source.visit(
+        [](GC::Root<HTMLImageElement> const& source) -> CSS::ImageOrientation {
+            auto const& computed_properties = source->computed_properties();
+            return computed_properties
+                ? computed_properties->image_orientation()
+                : CSS::ImageOrientation::FromImage;
+        },
+        [](GC::Root<SVG::SVGImageElement> const& source) -> CSS::ImageOrientation {
+            auto const& computed_properties = source->computed_properties();
+            return computed_properties
+                ? computed_properties->image_orientation()
+                : CSS::ImageOrientation::FromImage;
+        },
+        [](GC::Root<HTMLCanvasElement> const& source) -> CSS::ImageOrientation {
+            auto const& computed_properties = source->computed_properties();
+            return computed_properties
+                ? computed_properties->image_orientation()
+                : CSS::ImageOrientation::FromImage;
+        },
+        [](GC::Root<HTMLVideoElement> const& source) -> CSS::ImageOrientation {
+            auto const& computed_properties = source->computed_properties();
+            return computed_properties
+                ? computed_properties->image_orientation()
+                : CSS::ImageOrientation::FromImage;
+        },
+        [](GC::Root<ImageBitmap> const&) -> CSS::ImageOrientation { return CSS::ImageOrientation::FromImage; });
```
As most of these are the same code, a clever trick you can use is to put `auto const&` as the parameter (as long as it's after all the ones with explicit types). eg:

```suggestion
    auto const image_orientation = source.visit(
        [](GC::Root<ImageBitmap> const&) -> CSS::ImageOrientation { return CSS::ImageOrientation::FromImage; },
        [](auto const& source) -> CSS::ImageOrientation {
            auto const& computed_properties = source->computed_properties();
            return computed_properties
                ? computed_properties->image_orientation()
                : CSS::ImageOrientation::FromImage;
        });
```

AtkinsSJ commented on Libraries/LibWeb/HTML/DecodedImageData.h:
```diff
@@ -7,10 +7,15 @@
 #include <LibJS/Heap/Cell.h>
 #include <LibWeb/PixelUnits.h>
 
+namespace Gfx {
+enum class ImageOrientation : u8;
+}
+
```
Again, including the header should be fine.

AtkinsSJ commented on Libraries/LibWeb/HTML/HTMLImageElement.cpp:
```diff
@@ -171,22 +172,40 @@ bool HTMLImageElement::is_image_available() const
-    if (auto image_data = m_current_request->image_data())
-        return image_data->intrinsic_height();
+    if (auto image_data = m_current_request->image_data()) {
+        auto const image_orientation = computed_properties()
+            ? to_gfx_image_orientation(computed_properties()->image_orientation())
+            : Gfx::ImageOrientation::FromExif;
```
You have to do this a lot. Is it worth storing the Gfx::ImageOrientation in the computed properties instead of the CSS::ImageOrientation?

AtkinsSJ commented on Libraries/LibWeb/Painting/CanvasPaintable.cpp:
```diff
@@ -4,6 +4,7 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <LibWeb/CSS/ToGfxConversions.h>
```
This file doesn't seem to use this include.

AtkinsSJ reviewed (approved):
I think this is in a good enough shape to go ahead and merge it and save it from the endless conflicts. :sweat_smile: We can iterate on it once it's in tree and you're not stuck in "new contributor" mode. Thank you for your patience!

JohnyTheCarrot commented:
Same thing. My meds running out both at home and at pharmacies has been disastrous and has resulted in me having to rush to catch up a metric butt load of work. My apologies for the delay. As for why I haven't just quickly fixed this last failing test, I remember it being very much non-trivial to fix. Not something I can spare time for at this time. Either way, the deadlines on the stuff I'm catching up on are the 24th of this month and the 28th of this month respectively.

Bear with me.

JohnyTheCarrot committed:
LibGfx+LibWeb: Move EXIF orientation to rendering

JohnyTheCarrot commented:
Update: I will be applying my changes onto the latest commits from master manually, rather than toying with rebasing. It also ensures I double check everything. My current code will be backed up to a new branch on my end, and this PR's branch will be reset to upstream master, as I superimpose my changes gradually.

JohnyTheCarrot committed:
wip: second commit
